sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/InferVariantShreddingSchema.scala

```diff
@@ -0,0 +1,380 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.parquet
+
+import org.apache.spark.SparkRuntimeException
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types._
+import org.apache.spark.types.variant._
+import org.apache.spark.types.variant.VariantUtil.Type
+import org.apache.spark.unsafe.types._
+
+/**
+ *
+ * Infer a schema when there are Variant values in the shredding schema.
+ * Only VariantType values at the top level or nested in struct fields are replaced.
+ * VariantType nested in arrays or maps are not modified.
+ * @param schema The original schema containing VariantType.
+ */
+class InferVariantShreddingSchema(val schema: StructType) {
+
+  /**
+   * Create a list of paths to Variant values in the schema.
+   * Variant fields nested in arrays or maps are not included.
+   * For example, if the schema is
+   * struct<v: variant, struct<a: int, b: int, c: variant>>
+   * the function will return [[0], [1, 2]
+   */
+  private def getPathsToVariant(s: StructType): Seq[Seq[Int]] = {
+    s.fields.zipWithIndex
+      .map {
+        case (field, idx) =>
+          field.dataType match {
+            case VariantType =>
+              Seq(Seq(idx))
+            case inner: StructType =>
+              // Prepend this index to each downstream path.
+              getPathsToVariant(inner).map { path =>
+                idx +: path
+              }
+            case _ => Seq()
+          }
+      }
+      .toSeq
+      .flatten
+  }
+
+  private def getValueAtPath(s: StructType, row: InternalRow, p: Seq[Int]): Option[VariantVal] = {
```

Comment:
NIT:
```suggestion
  private def getValueAtPath(schema: StructType, row: InternalRow, path: Seq[Int]): Option[VariantVal] = {
```

Also, can we comment on what this will return? It looks it will return `None` of the field will be null?